File: backend/app/routes/resume.py
import os
import shutil
from fastapi import APIRouter, Depends, HTTPException, UploadFile, File
from datetime import datetime
from app.database.db_manager import get_db, get_next_sequence_value
from app.models.user import User
from app.models.resume import Resume
from app.routes.auth import get_current_user
from app.config import Config
from app.utils.file_validator import validate_file

router = APIRouter(prefix="/api/resume", tags=["resume management"])

@router.post("/upload")
async def upload_resume(
    file: UploadFile = File(...),
    current_user: User = Depends(get_current_user),
    db = Depends(get_db)
):
    # Validate file
    validation_result = validate_file(file, Config.ALLOWED_EXTENSIONS, Config.MAX_FILE_SIZE)
    if not validation_result["valid"]:
        raise HTTPException(status_code=400, detail=validation_result["error"])
    
    # Create upload directory if not exists
    os.makedirs(Config.UPLOAD_DIR, exist_ok=True)
    
    # Generate unique filename
    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
    file_extension = os.path.splitext(file.filename)[1]
    unique_filename = f"{current_user.user_id}_{timestamp}{file_extension}"
    file_path = os.path.join(Config.UPLOAD_DIR, unique_filename)
    
    # Save file
    with open(file_path, "wb") as buffer:
        shutil.copyfileobj(file.file, buffer)
    
    # Save to database
    resume_id = await get_next_sequence_value("resume_id")
    resume_doc = {
        "resume_id": resume_id,
        "user_id": current_user.user_id,
        "resume_file_name": file.filename,
        "file_format": file_extension,
        "file_size": file.size,
        "upload_date": datetime.utcnow(),
        "resume_path": file_path
    }
    await db["resumes"].insert_one(resume_doc)
    
    return {
        "message": "Resume uploaded successfully",
        "resume_id": resume_id,
        "file_name": file.filename
    }

@router.get("/list")
async def get_user_resumes(
    current_user: User = Depends(get_current_user),
    db = Depends(get_db)
):
    resumes_cursor = db["resumes"].find({"user_id": current_user.user_id})
    resumes = await resumes_cursor.to_list(None)
    return [
        {
            "resume_id": r["resume_id"],
            "file_name": r["resume_file_name"],
            "upload_date": r["upload_date"],
            "file_size": r["file_size"]
        }
        for r in resumes
    ]

File: backend/src/routes/resume.js
const express = require('express');
const path = require('path');
const fs = require('fs');
const multer = require('multer');
const { getDB, getNextSequenceValue } = require('../db');
const { createResumeDoc } = require('../models/Resume');
const authenticate = require('../middleware/auth');
const config = require('../config');

const router = express.Router();

// ─── Multer (file upload) ─────────────────────────────────────────────────────

// Ensure upload directory exists
fs.mkdirSync(config.UPLOAD_DIR, { recursive: true });

const storage = multer.diskStorage({
  destination: (_req, _file, cb) => cb(null, config.UPLOAD_DIR),
  filename: (req, file, cb) => {
    const timestamp = new Date().toISOString().replace(/[-:.TZ]/g, '').slice(0, 15);
    const ext = path.extname(file.originalname).toLowerCase();
    cb(null, `${req.user.user_id}_${timestamp}${ext}`);
  },
});

const fileFilter = (_req, file, cb) => {
  const ext = path.extname(file.originalname).toLowerCase();
  if (config.ALLOWED_EXTENSIONS.includes(ext)) {
    cb(null, true);
  } else {
    // In multer 2.x passing an Error as first arg still works
    cb(new multer.MulterError('LIMIT_UNEXPECTED_FILE', `Invalid file format. Allowed formats: ${config.ALLOWED_EXTENSIONS.join(', ')}`));
  }
};

const upload = multer({
  storage,
  fileFilter,
  limits: { fileSize: config.MAX_FILE_SIZE },
});

// Upload middleware wrapped to convert multer errors into JSON responses
function uploadSingle(req, res, next) {
  upload.single('file')(req, res, (err) => {
    if (!err) return next();
    if (err instanceof multer.MulterError) {
      if (err.code === 'LIMIT_FILE_SIZE') {
        return res.status(400).json({ detail: `File too large. Maximum size: ${config.MAX_FILE_SIZE / (1024 * 1024)}MB` });
      }
      return res.status(400).json({ detail: err.field || err.message });
    }
    return res.status(400).json({ detail: err.message });
  });
}

// ─── POST /api/resume/upload ──────────────────────────────────────────────────
router.post('/upload', authenticate, uploadSingle, async (req, res) => {
  if (!req.file) {
    return res.status(400).json({ detail: 'No file uploaded' });
  }

  try {
    const db = getDB();
    const file = req.file;
    const ext = path.extname(file.originalname).toLowerCase();

    const resumeId = await getNextSequenceValue('resume_id');
    const resumeDoc = createResumeDoc({
      resume_id: resumeId,
      user_id: req.user.user_id,
      resume_file_name: file.originalname,
      file_format: ext,
      file_size: file.size,
      upload_date: new Date(),
      resume_path: file.path,
    });

    await db.collection('resumes').insertOne(resumeDoc);

    return res.status(201).json({
      message: 'Resume uploaded successfully',
      resume_id: resumeId,
      file_name: file.originalname,
    });
  } catch (dbErr) {
    console.error('Resume upload DB error:', dbErr);
    return res.status(500).json({ detail: `Upload failed: ${dbErr.message}` });
  }
});

// ─── GET /api/resume/list ─────────────────────────────────────────────────────
router.get('/list', authenticate, async (req, res) => {
  try {
    const db = getDB();
    const resumes = await db
      .collection('resumes')
      .find({ user_id: req.user.user_id })
      .toArray();

    return res.json(
      resumes.map((r) => ({
        resume_id: r.resume_id,
        file_name: r.resume_file_name,
        upload_date: r.upload_date,
        file_size: r.file_size,
      }))
    );
  } catch (err) {
    console.error('List resumes error:', err);
    return res.status(500).json({ detail: err.message });
  }
});

module.exports = router;
